monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/Interleave2Observable.scala

```diff
@@ -0,0 +1,215 @@
+            case Stop     ⇒ self.synchronized(if (select == 0) rawOnNext(a) else Stop)
+          }
+      }
+
+      continueA2.tryCompleteWith(lastAck)
+      continueA2 = Promise[Ack]()
+      lastAck
+    }
+
+    def signalOnError(ex: Throwable): Unit = self.synchronized {
+      if (!isDone) {
+        isDone = true
+        out.onError(ex)
+        lastAck = Stop
+        continueA1.tryCompleteWith(Stop)
+        continueA2.tryCompleteWith(Stop)
+      }
+    }
+
+    def signalOnComplete(ack: Future[Ack]): Unit = {
+      @inline def rawOnComplete(): Unit =
+        if (!isDone) {
+          isDone = true
+          out.onComplete()
+        }
+
+      self.synchronized  {
+        val shouldComplete = !isDone && (
+          {
+            completedCount += 1
+            completedCount == 2
+          })
+
+        if (shouldComplete) {
+          ack match {
+            case Continue ⇒ rawOnComplete()
+            case Stop     ⇒ () // do nothing
+            case async    ⇒
+              async.onComplete {
+                case Success(Continue) ⇒
+                  self.synchronized(rawOnComplete())
+                case _                 ⇒
+                  () // do nothing
+              }
+          }
+
+          lastAck = Stop
+        }
+      }
+    }
+
+    val composite = CompositeCancelable()
+
+    composite += obsA1.unsafeSubscribeFn(new Subscriber[A] {
+      implicit val scheduler = out.scheduler
+
+      def onNext(elem: A): Future[Ack] = self.synchronized {
+        if (isDone) Stop else {
+          hasElemA1 = true
+          if (select == 0) select = 1
+
+          lastAck1 = if (select == 1)
+            signalOnNextA1(elem)
+          else
+          // here essentially means waiting until obsA2 is done with its work
+          // when that happens, before passing the `Continue` signal to the upstream observable,
+          // immediately signal `onNext` of the out Observable to emit the elem since it is already available
+            continueA2.future.flatMap {
+              case Continue ⇒
+                self.synchronized { signalOnNextA1(elem) }
+              case Stop     ⇒
+                self.synchronized {
+                  if (select >= 0)
+                    signalOnNextA1(elem)
+                  else Stop
+                }
+            }
+
+          lastAck1
+        }
+      }
+
+      def onError(ex: Throwable): Unit =
+        signalOnError(ex)
+      def onComplete(): Unit = self.synchronized {
+        signalOnComplete(lastAck1)
+        continueA1.tryCompleteWith(Stop)
+        select = 0
+      }
+    })
+
+    composite += obsA2.unsafeSubscribeFn(new Subscriber[A] {
+      implicit val scheduler = out.scheduler
+
+      def onNext(elem: A): Future[Ack] = self.synchronized {
+        if (isDone) Stop else {
+          hasElemA2 = true
+          if (select == 0) select = -1
+
+          lastAck2 = if (select == -1)
+            signalOnNextA2(elem)
```

Comment:
OK, I realized I don't like something in your logic :-)

You're using `select` to decide which should emit. This is not right, for one because it is more complicated and because the `continueA1` and `continueA2` promises are then superfluous.

IMHO, this `onNext` method should start like this:

```scala
def onNext(elem: A): Future[Ack] = self.synchronized {
  continueA1.future.flatMap {
    case Continue => ???
    case Stop => ???
  }
}
```

In other words, the permission to signal `A2` is given entirely by `continueA1` and nothing else. And `continueA2` should start as an already completed promise, such that `A1` can be signaled at first.